### PR TITLE
feat(web): add ?strict=true mode to /api/health/ready for empty-shard detection

### LINKED ISSUE #1511: [FR] Add ?strict=true mode to /api/health/ready for empty-shard detection
## Problem

`GET /api/health/ready` (added in sourcebot-dev/sourcebot#1506 / PR [#1507](<https://github.com/sourcebot-dev/sourcebot/issues/1507>)) probes Postgres, Redis, and Zoekt and returns `503 / status: "degraded"` if any check fails. The Zoekt check uses an empty `List` RPC, which the gRPC server treats as a successful round-trip **even when no repositories are indexed**. That is the right default — the question a readiness probe answers is "can this instance talk to its dependencies?" — but it leaves a real operational gap: a freshly-restarted Sourcebot that has not yet completed a single sync, or a Sourcebot whose connection sync is silently broken, still reports `status: "ok"` and gets routed traffic.

## Motivation

A self-hosted operator wiring Sourcebot into a Kubernetes `readinessProbe` (or a load-balancer health check) typically wants two distinct signals:

* **Soft ready** — the process is up and the dependencies are reachable. The current default.
* **Strictly ready** — the process is up, the dependencies are reachable, and the instance can actually serve a search that returns results. A node that has not finished indexing anything is *not* strictly ready, even if all three RPCs succeed.

Today there is no way to ask the readiness probe the second question. A node can pass liveness, pass readiness, and still be useless to a search request.

## Proposed solution

Add a `?strict=true` query parameter to `GET /api/health/ready`. When `strict=true`:

* The Zoekt check inspects the `List` response. If it is empty (`repos === undefined` or `repos.length === 0`), the check is reported as `status: "degraded"` (or a new `status: "empty"` value), and the overall response becomes `503 / status: "degraded"`.
* The Postgres and Redis checks behave identically to the non-strict path.
* The response shape gains a `strict: true | false` field so callers can confirm which mode was applied.

When `strict` is absent or `strict=false` (the default), behavior is unchanged from sourcebot-dev/sourcebot#1506.

## Example

```
$ curl -s 'http://localhost:3000/api/health/ready?strict=true' | jq
{
  "status": "degraded",
  "strict": true,
  "checks": {
    "postgres": { "status": "ok", "latencyMs": 3 },
    "redis":    { "status": "ok", "latencyMs": 1 },
    "zoekt":    { "status": "degraded", "latencyMs": 18, "error": "no repositories indexed" }
  }
}
```

## Alternatives considered

* **Add a separate** `/api/health/ready/strict` **endpoint.** Rejected: duplicate route, duplicate auth, duplicate PostHog tracking, no real benefit over a query param.
* **Make** `strict` **the default and add a** `?loose=true` **opt-out.** Rejected: a backwards-incompatible behavior change for a recently-merged endpoint is a poor first move; the loose interpretation is the natural one for "can we talk to our dependencies".
* **Caching the response and using the cached value to drive the strict signal.** Out of scope. Caching is a separate, larger follow-on already noted in sourcebot-dev/sourcebot#1506's "future work" section.

## Scope

* New query parameter on the existing route.
* Per-check shape gains a `strict`-relevant error string for the Zoekt case.
* New tests covering: strict with empty shards → 503; strict with shards → 200; non-strict with empty shards → 200; invalid `strict` value → 400.
* Docs (`docs/docs/api-reference/health.mdx`) updated with one example showing the strict path and a Kubernetes `readinessProbe` snippet that uses it.
* No new dependencies, no schema migration, no new files. Pure addition to the existing route.

## Acceptance criteria

- [ ] `GET /api/health/ready?strict=true` returns `503 / status: "degraded"` when Zoekt has zero indexed repos, and `200 / status: "ok"` when Zoekt has at least one.
- [ ] `GET /api/health/ready` (no query param) and `?strict=false` keep their current behavior.
- [ ] `?strict=invalid` returns `400` with a clear error message.
- [ ] The response always includes the `strict` field so operators can confirm the mode.
- [ ] New tests cover all four cases above plus the parallel-execution invariant from sourcebot-dev/sourcebot#1506.
- [ ] Lint, type-check, and full test pass on the modified files.
- [ ] Docs updated; `CHANGELOG.md` entry added under `[Unreleased] → Added`.

## Backward compatibility

Fully backwards compatible. The default behavior is unchanged. Operators who do not pass `strict=true` see the same response as today.

## Risks

* The Zoekt `List` response shape differs between proto versions. Need to handle both `repos` (RepoListFieldRepos) and `reposMap` (RepoListFieldReposMap) — empty in both cases means strict-not-ready.
* Operators using a default `?strict=true` probe who do not see any indexed repos in their dev environment may be confused why readiness fails. Mitigated by the `strict` field in the response and a clear error string on the Zoekt check.

## Related

* Closes the "strict=true mode" follow-on item in the "Out of scope" section of sourcebot-dev/sourcebot#1506.
* Builds on `GET /api/health/ready` from sourcebot-dev/sourcebot#1506 / PR [#1507](<https://github.com/sourcebot-dev/sourcebot/issues/1507>).

### PR DESCRIPTION
**Pull request by Harsh23Kashyap**
## Summary

Adds `?strict=true` to `GET /api/health/ready` for empty-shard detection. The default behavior of the readiness probe is unchanged: a Zoekt `List` RPC that returns an empty result still counts as a successful probe, because the question a readiness probe answers is "can this instance talk to its dependencies?". With `?strict=true`, an empty Zoekt shard set is treated as `503 / status: "degraded"` and `zoekt.status: "empty"`, distinguishing "Zoekt is up but the instance has not finished indexing" from "Zoekt is unreachable".

Fixes #1511.

## Motivation

A self-hosted operator wiring Sourcebot into a Kubernetes `readinessProbe` (or a load-balancer health check) typically wants two distinct signals:

- **Soft ready** — the process is up and the dependencies are reachable. The default mode.
- **Strictly ready** — the process is up, the dependencies are reachable, and the instance can actually serve a search that returns results. A node that has not finished indexing anything is *not* strictly ready, even if all three RPCs succeed.

Today there is no way to ask the probe the second question. A node can pass liveness, pass readiness, and still be useless to a search request. The default behavior is the right one for "can we talk to our dependencies?" but operators who want "is this instance useful?" need a way to ask it.

## Changes

- **`packages/web/src/app/api/(server)/health/ready/route.ts`** — adds the `strict` query param (parsed by a Zod schema; rejects anything other than the literal strings `"true"` / `"false"` as 400), threads it into `checkZoekt`, and adds a third check status `"empty"` for the strict-empty case. The `checkZoekt` function now actually captures and inspects the `List` response (it previously only consumed the callback's error argument). The gRPC callback signature in `zoektClient.ts` is corrected accordingly: it always was `(err, result) => void`, not `(err) => void`.
- **`packages/web/src/lib/zoektClient.ts`** — adds a loose `ZoektListResponse` summary type alongside the existing `ZoektClient` so the route and its test can refer to the same shape without pulling in the full gRPC response type. Fixes the `List` callback signature.
- **`packages/web/src/app/api/(server)/health/ready/route.test.ts`** — seven new test cases: strict with one indexed repo → 200; strict with empty `repos` → 503 / `empty`; strict with empty `repos_map` (the `RepoListFieldReposMap` field) → 503 / `empty`; strict with non-empty `repos_map` → 200; non-strict with empty `repos` → 200 (backward-compat); absent `strict` parameter treated as `false`; invalid `strict` value → 400. Existing tests updated to pass a minimal `NextRequest` stand-in (the route now needs `nextUrl.searchParams`).
- **`docs/docs/api-reference/health.mdx`** — new "Strict mode" subsection explaining the operator motivation, the new response fields, and the `error` vs `empty` distinction. The K8s `readinessProbe` example gets a comment pointing at the strict path for deployments that should not receive traffic until the shard set is non-empty.
- **`CHANGELOG.md`** — Added entry under `[Unreleased]` linking to #1511.

## Design decisions

- **Query param, not separate endpoint.** A separate `/api/health/ready/strict` would duplicate the route, the auth policy, the PostHog tracking opt-out, and the per-check response shape. The query param keeps it additive.
- **Default unchanged.** Operators who do not pass `?strict=true` see exactly the same response as before. A backwards-incompatible behavior change for a recently-merged endpoint is a poor first move.
- **Third status `"empty"` instead of overloading `"error"`.** `"error"` means the check itself failed (Zoekt unreachable, timeout, etc.). `"empty"` means the check passed but the result is empty in strict mode. They are different operational signals and an operator looking at the response should be able to tell them apart without parsing the `error` string.
- **Check both `repos` and `repos_map`.** The Zoekt gRPC server populates different fields depending on `ListOptions.Field` (`RepoListFieldRepos` vs `RepoListFieldReposMap`). An empty response in either shape is "strictly empty" and the check covers both.
- **`z.string().refine(...)` instead of `z.coerce.boolean()`.** The latter is just `Boolean(value)` under the hood, which would treat the string `"false"` as truthy. Accept only the two literal strings we mean to support, refuse anything else as 400.
- **New `ZoektListResponse` summary type.** The full gRPC response type is much wider than we need; exporting a narrow summary keeps the test mock small and avoids coupling the route's error handling to the full proto type.

## Verification

- `yarn workspace @sourcebot/web test --run "health/ready"` — 15/15 tests pass
- `yarn workspace @sourcebot/web lint` — 0 errors (4 pre-existing warnings in unrelated files)
- `tsc --noEmit -p packages/web/tsconfig.json` — 0 errors in the new files

## Backward compatibility

Fully backwards compatible. The default response shape gains a single new field (`strict`); all other fields and all status codes are unchanged when `strict` is absent or `false`.

## Stack

This PR is stacked on #1507 (the readiness probe itself). The first commit in the diff (`f5193aea`) is the `max_wall_time` follow-up from the PR #1507 review, not new work for this PR. Once #1507 merges, this branch will be rebased onto `main` and that commit will fall away; the remaining four commits are the actual `?strict=true` work.

## Out of scope (tracked in #1511 as future work)

- Caching the readiness result for ~1s to absorb thundering-herd probes from multiple orchestrators.
- Per-check Prometheus metrics for the strict case.

## Risks

- The Zoekt `List` response shape differs between proto versions. The check handles both `repos` and `repos_map`; an empty result in either is treated as strict-empty. The existing `zoektClient` returns the full gRPC `__Output` shape, so no information is lost.
- Operators using `?strict=true` in a dev environment with no indexed repos will see the probe fail with a clear `error: "no repositories indexed (strict mode)"` rather than a generic "degraded". The docs call this out.

---

> [!NOTE]
> **Low Risk**
> Additive, unauthenticated operational endpoints with backward-compatible defaults; main caveat is probe load on Postgres and misconfigured strict readiness blocking traffic until indexing completes.
> 
> **Overview**
> Adds **`GET /api/health/ready`**, a public readiness probe that checks Postgres (`SELECT 1`), Redis (`PING`), and Zoekt (empty `List` gRPC) in parallel with a 2s per-check timeout. It returns **200** with per-check latency when all pass, or **503** with `status: "degraded"` and generic client-safe errors (details logged server-side). **`GET /api/health`** stays the lightweight liveness probe.
> 
> Optional **`?strict=true`** treats an empty Zoekt shard set as not ready: **`zoekt.status: "empty"`** and **503**, separate from **`error`** when Zoekt is unreachable. Default / `strict=false` behavior is unchanged (empty index still counts as reachable). Invalid `strict` values return **400**.
> 
> Introduces **`packages/web/src/lib/zoektClient.ts`** (mockable gRPC client + `ZoektListResponse`) and extensive Vitest coverage. Docs add **`docs/api-reference/health.mdx`** (K8s/Docker examples, strict mode); **`CHANGELOG.md`** and **`docs.json`** are updated.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 44e43bfe0584416c644d1c5eb1df9a48a70b7ea1. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **New Features**
  * Added health endpoints: liveness (`GET /api/health`) and readiness (`GET /api/health/ready`) that probe Postgres, Redis, and Zoekt.
  * Readiness returns detailed per-service check results (including latency/errors) and supports `?strict=true` to mark health degraded when Zoekt has no indexed repositories.
* **Documentation**
  * Added API reference docs for health endpoints and updated API navigation.
  * Updated changelog for the new readiness endpoint and strict-mode behavior.
* **Bug Fixes**
  * Improved readiness validation and error handling to avoid leaking internal details.
* **Tests**
  * Expanded automated coverage for success/degraded/strict scenarios and edge cases.

### CODE REVIEW

File: packages/web/src/app/api/(server)/health/ready/route.ts [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+                });
+            });
+        }, READINESS_TIMEOUT_MS);
+
+        if (strict) {
+            // Check both `repos` (Field = RepoListFieldRepos) and
+            // `repos_map` (Field = RepoListFieldReposMap) so the result is
+            // correct regardless of which field the server populates.
+            const repos = response.repos ?? [];
+            const reposMap = response.repos_map ?? {};
```
[the remark this answers is not in the document]
Comment (Harsh23Kashyap) [reply packages/web/src/app/api/(server)/health/ready/route.ts:167]: Good catch — fixed in 3fa3c63c.

Confirmed: the gRPC callback can fire with `err === null` and `result === undefined` on a malformed response. Without a guard, the strict path would throw on `response.repos` and surface the failure as a generic Node unhandled-rejection path instead of the controlled `error` status that the response shape advertises.

`zoektSearch` applies the same `error || !response` pattern (`features/search/zoektSearcher.ts`). Mirrored it in the readiness route:

```ts
client.List({}, (err, result) => {
    if (err) {
        reject(err);
    } else if (!result) {
        reject(new Error('zoekt List RPC returned no response'));
    } else {
        resolve(result);
    }
});
```

Added a regression test that fires the List callback with `callback(null, undefined)` and asserts the response carries `zoekt.status:"error"` and an `error` string mentioning the missing response.

16/16 tests pass; lint 0 errors; tsc 0 errors in the new files.

File: packages/web/src/lib/zoektClient.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    const proto = grpc.loadPackageDefinition(packageDefinition) as unknown as {
+        zoekt: { webserver: { v1: { WebserverService: new (address: string, credentials: unknown) => ZoektClient } } };
+    };
+
+    const zoektUrl = new URL(shared.env.ZOEKT_WEBSERVER_URL);
+    const grpcAddress = `${zoektUrl.hostname}:${zoektUrl.port}`;
+    return new proto.zoekt.webserver.v1.WebserverService(
+        grpcAddress,
+        grpc.credentials.createInsecure(),
+    );
```
[the remark this answers is not in the document]
Comment (Harsh23Kashyap) [reply packages/web/src/lib/zoektClient.ts:60]: Good catch — fixed in a92eceb0.

Confirmed: `zoektSearcher` sets 500MB message-size limits on the gRPC channel; the readiness probe's `loadZoektClient` did not. A healthy Zoekt instance with a large repo catalog (or a stretched deployment) could hit the default 4MB receive cap, fail the probe, and report `zoekt.status:"error"` even though search against the same backend would succeed. The readiness route is the wrong place for that failure to surface.

```ts
return new proto.zoekt.webserver.v1.WebserverService(
    grpcAddress,
    grpc.credentials.createInsecure(),
    {
        'grpc.max_receive_message_length': 500 * 1024 * 1024, // 500MB
        'grpc.max_send_message_length':    500 * 1024 * 1024, // 500MB
    },
);
```

Also loosened the loose constructor type in `zoektClient.ts` to accept the third `options` argument. The old type was hard-coded to two parameters, which would have rejected this change at tsc time.

16/16 tests pass; lint 0 errors; tsc 0 errors in the new files.

File: CHANGELOG.md [hunk 3]
```diff
@@ -7,6 +7,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Added
+- Added a `GET /api/health/ready` endpoint that returns per-dependency health (Postgres, Redis, Zoekt) for use as a Kubernetes `readinessProbe` or load-balancer health check. The existing `GET /api/health` endpoint is unchanged and remains the liveness probe. [#1507](https://github.com/sourcebot-dev/sourcebot/pull/1507)
+- Added a `?strict=true` query parameter to `GET /api/health/ready`. When set, the endpoint reports `503 / status: "degraded"` and `zoekt.status: "empty"` if the Zoekt shard set contains no indexed repositories, distinguishing "Zoekt is unreachable" from "Zoekt is up but the instance is not yet useful" for orchestrators. Default behavior is unchanged. [#1511](https://github.com/sourcebot-dev/sourcebot/issues/1511)
```
[the remark this answers is not in the document]
Comment (Harsh23Kashyap) [reply CHANGELOG.md]: Good catch — fixed in 44e43bfe. Changelog entry now points at #1512 (the PR), not #1511 (the issue).

File: docs/docs.json [hunk 4]
```diff
@@ -217,7 +217,8 @@
                         "icon": "server",
                         "pages": [
                             "GET /api/version",
-                            "GET /api/health"
+                            "GET /api/health",
+                            "docs/api-reference/health"
                         ]
```
[the remark this answers is not in the document]
Comment (Harsh23Kashyap) [reply docs/docs.json:221]: Good catch — fixed in 2b8b833d.

Confirmed: `docs/api-reference/health.mdx` already documents both `/api/health` (Liveness section) and `/api/health/ready` (Readiness section), so the mdx page is the canonical doc for the health endpoints. The standalone `GET /api/health` entry in the System group was sourced from the OpenAPI document and created a second nav item pointing at the same surface.

Removed the `GET /api/health` line. `docs/api-reference/health` is now the single nav entry for both endpoints.

(Adding the new endpoint to the OpenAPI spec would be a larger change touching the public API spec regeneration; the mdx page is sufficient for the docs surface and keeps the new endpoint as a focused addition.)

File: packages/web/src/app/api/(server)/health/ready/route.ts [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+        }, READINESS_TIMEOUT_MS);
+        return { status: 'ok', latencyMs: Date.now() - start };
+    } catch (err) {
+        return {
+            status: 'error',
+            latencyMs: Date.now() - start,
+            error: err instanceof Error ? err.message : String(err),
+        };
+    }
+};
```
[the remark this answers is not in the document]
Comment (Harsh23Kashyap) [reply packages/web/src/app/api/(server)/health/ready/route.ts:104]: Good catch — fixed in 4ef628d1.

Confirmed: `checkPostgres`, `checkRedis`, and `checkZoekt` were returning raw `err.message` in the public JSON response. Since the route is unauthenticated and may be reachable at a public URL, those messages can leak internal hostnames, ports, or connection details to any caller (Prisma's connection-refused error includes the host:port it tried; the gRPC UNAVAILABLE error includes the target address).

Changes:
- `CheckResult` now has a server-side-only `errorDetail` field that carries the raw `err.message`.
- The public `error` field on degraded checks is a generic `"<label> check failed: see server logs"` string.
- The GET handler logs the FULL per-check result (including `errorDetail`) when the overall probe is degraded, so operators still have the diagnostic detail on the server side.
- The response body uses a stripped view that omits `errorDetail`.

Tests updated: the four degraded-path tests now assert the generic public error string, that the internal message (containing fake hostnames like `db.internal.example.com`, `10.0.0.42:6379`, `zoekt.svc.cluster.local:6070`) does not appear in the response, that `errorDetail` is undefined on the wire, and that the full message is present in the server-side log call.

16/16 tests pass; lint 0 errors; tsc 0 errors in the new files.

**Comment by Harsh23Kashyap** [comment]
Opened #1512 to add `?strict=true` to `/api/health/ready`. Default behavior is unchanged; strict mode additionally probes the Zoekt shard list and returns 503 if the list is empty, even when the existing zoekt call would have returned 200 (empty gRPC List response is success on the wire but indicates no shards registered). Useful for distinguishing 'zoekt is down' from 'zoekt is up but no repos indexed' in orchestrator probes. CI pending the vulnerability-triage workflow.
